infrastructure/json/src/main/java/tech/pegasys/teku/infrastructure/json/types/SerializableOneOfTypeDefinition.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2022 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.infrastructure.json.types;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.stream.Collectors.toSet;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+
+public class SerializableOneOfTypeDefinition<TObject>
+    implements SerializableTypeDefinition<TObject> {
+  private final Map<Class<? extends TObject>, SerializableTypeDefinition<? extends TObject>> types;
+  private final Optional<String> name;
+  private final Optional<String> title;
+  private final Optional<String> description;
+
+  SerializableOneOfTypeDefinition(
+      final Optional<String> name,
+      final Optional<String> title,
+      final Optional<String> description,
+      final Map<Class<? extends TObject>, SerializableTypeDefinition<? extends TObject>> types) {
+    this.name = name;
+    this.title = title;
+    this.description = description;
+    this.types = types;
+  }
+
+  @Override
+  public void serializeOpenApiType(final JsonGenerator gen) throws IOException {
+    gen.writeStartObject();
+    if (description.isPresent()) {
+      gen.writeStringField("description", description.get());
+    }
+    gen.writeStringField("type", "object");
+    gen.writeArrayFieldStart("oneOf");
+    for (SerializableTypeDefinition<? extends TObject> t : types.values()) {
+      t.serializeOpenApiTypeOrReference(gen);
+    }
+    gen.writeEndArray();
+    gen.writeEndObject();
+  }
+
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  @Override
+  public void serialize(TObject value, final JsonGenerator gen) throws IOException {
+    final SerializableTypeDefinition typeDefinition = types.get(value.getClass());
+    checkArgument(
+        typeDefinition != null, "No class serialization method found: %s", value.getClass());
+    typeDefinition.serialize(value, gen);
+  }
+
+  @Override
+  public SerializableTypeDefinition<TObject> withDescription(final String description) {
+    return new SerializableOneOfTypeDefinition<>(name, title, Optional.of(description), this.types);
```

Comment:
Remove the name when changing the description - otherwise we'll wind up with two different objects referenced by the same name.